refactor(config): migrate from `figment` to `config-rs`

gustavovalverde opened the pull request:
### Motivation
Unify configuration across the Z3 stack and harden handling of sensitive values. This PR migrates from `figment` to `config-rs` and prevents sensitive fields from being overridden via environment variables, aligning Zaino with Zebra and improving default security posture.

### Solution
- Replace `figment` with `config-rs` using a 3-layer hierarchy: defaults → TOML → env
- Add case-insensitive, suffix-based env deny-list with **explicit error messages**:
  - Block env overrides whose leaf key ends with: `password`, `secret`, `token`, `cookie`, `private_key`
  - Returns clear error explaining why and directing users to config file
  - Supports nested config via `ZAINO_` prefix with `__` separator (e.g., `ZAINO_VALIDATOR_SETTINGS__VALIDATOR_USER`)
- Add `#[serde(deny_unknown_fields)]` to catch typos in TOML config files
- Add `#[serde(default)]` to config structs for partial config support
- Comprehensive inline test coverage using `EnvGuard` pattern

### Changes
| File | Change |
|------|--------|
| `Cargo.toml` | `figment` → `config-rs` |
| `zainod/Cargo.toml` | Update dependencies |
| `zainod/src/config.rs` | Major rewrite with config-rs, sensitive key protection, inline tests |
| `zainod/src/indexer.rs` | Simplify BackendType matching |
| `zaino-state/src/config.rs` | Add `Default`, `Eq`, `Serialize` to `BackendType` |
| `zaino-serve/src/server/config.rs` | Add `#[serde(alias)]` for backwards compatibility |
| `zaino-common/src/config/storage.rs` | Add `#[serde(default)]` |
| `zaino-common/src/config/validator.rs` | Add `#[serde(default)]` and `impl Default` |

### Security Enhancement
This PR prevents accidental exposure of sensitive configuration values through environment variables, which is particularly important in containerized deployments where env vars might be logged or visible to other processes.

**Key difference from silent filtering**: Instead of silently dropping sensitive env vars, this implementation returns an explicit error message explaining why the env var is blocked and directing users to use the config file instead. This follows the "fail loudly" principle for better security UX.

**Blocked env overrides**: `ZAINO_VALIDATOR_SETTINGS__VALIDATOR_PASSWORD`, `ZAINO_*_SECRET`, `ZAINO_*_TOKEN`, etc.
**Allowed env overrides**: `ZAINO_NETWORK`, `ZAINO_GRPC_SETTINGS__LISTEN_ADDRESS`, etc.

### Testing
- 15 config tests pass (inline in `zainod/src/config.rs`)
- Tests cover: sensitive key blocking, env overrides, TOML parsing, unknown field rejection
- Backward compatibility maintained via `#[serde(alias)]` for existing config files

### Related
- Aligns with Zebra's config security model
- Part of broader Z3 stack configuration unification

y4ssi commented on zainod/src/config.rs:
```diff
@@ -20,6 +20,17 @@ use zaino_state::{BackendConfig, FetchServiceConfig, StateServiceConfig};
 
+/// Returns true if a leaf key name should be considered sensitive and blocked
+/// from environment variable overrides.
+fn is_sensitive_leaf_key(leaf_key: &str) -> bool {
```
Maybe it would be good to add a sensitive_ prefix, so that over time we don’t have to maintain the variables one by one under this condition, and we could establish it across the Z3 stack.

gustavovalverde replied:
Solved

gustavovalverde committed:
refactor(config): migrate from figment to config-rs

Replace figment with config-rs for configuration management, following
the 3-layer hierarchy: defaults → TOML → environment variables.

Key changes:
- Add sensitive key protection that errors on env vars ending with
  password, secret, token, cookie, or private_key
- Support nested config via ZAINO_ prefix with __ separator
- Add #[serde(deny_unknown_fields)] for strict TOML validation
- Add #[serde(default)] to structs for partial config support
- Add serde alias for backwards compatibility with grpc_listen_address
- Simplify BackendType matching in indexer (remove BackendConfig wrapper)
- Convert tests from Figment Jail to EnvGuard pattern

Security: Environment variables for sensitive fields are blocked with
clear error messages, requiring use of config files instead.

gustavovalverde committed:
docs(config): fix outdated comment about cookie auth requirement

The comment incorrectly stated cookie auth was required for non-localhost
addresses. Update to reflect actual behavior: authentication is recommended
but not enforced.
